README.md
# replayman

Log agent to replay time-stamped log stream.

## Getting started

### Installation

```bash
cargo install replayman
```

### Prepare for your data

TSV with a header line is supported.

```bash
curl -O https://raw.githubusercontent.com/SpringQL/dataset/main/pseudo-in-vehicle/AirConditioner-30sec.tsv
```

### Usage

#### stdout

```bash
$ replayman \
  --timed-by Time \
  --initial-timestamp '2020-10-21T10:37:56.000+09:00' \
  AirConditioner-30sec.tsv
```

#### TCP socket

```bash
nc -l 19870
```

```bash
$ replayman \
  --timed-by Time \
  --initial-timestamp '2020-10-21T10:37:56.000+09:00' \
  --dest-tcp '127.0.0.1:19870' \
  AirConditioner-30sec.tsv
```

#### MQTT

Using [mosquitto](https://github.com/eclipse/mosquitto) here for example.

```bash
mosquitto_sub -h 127.0.0.1 -t replayman/test
```

```bash
$ replayman \
  --timed-by Time \
  --initial-timestamp '2020-10-21T10:37:56.000+09:00' \
  --dest-mqtt '127.0.0.1:1883' \
  --dest-mqtt-topic 'replayman/test' \
  AirConditioner-30sec.tsv
```

#### Kafka

```bash
# create `replayman-test` topic
kafka-topics --create --bootstrap-server localhost:9092 --topic replayman-test --replication-factor 1 --partitions 1
# subscribe to `replayman-test` topic
kafka-console-consumer --bootstrap-server localhost:9092 --topic replayman-test --from-beginning
```

```bash
$ replayman \
  --timed-by Time \
  --initial-timestamp '2020-10-21T10:37:56.000+09:00' \
  --dest-kafka-bootstrap '127.0.0.1:9092' \
  --dest-kafka-topic 'replayman-test' \
  AirConditioner-30sec.tsv
```

## License

Licensed under either of [Apache License, Version 2.0](LICENSE-APACHE) or [MIT license](LICENSE-MIT) at your option.

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in replayman by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.

Copyright (c) 2021-2022 TOYOTA MOTOR CORPORATION.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][Keep a Changelog] and this project adheres to [Semantic Versioning][Semantic Versioning].

<!-- markdownlint-disable MD024 -->
## [Unreleased]

## [v0.5.1] - 2022-11-09

(Internal changes only)

## [v0.5.0] - 2022-11-09

### Added

- log to Kafka [#85](https://github.com/SpringQL/replayman/pull/85)

## [v0.4.0] - 2022-11-06

### Added

- log to stdout [#81](https://github.com/SpringQL/replayman/pull/81)

## [v0.3.0] - 2022-11-04

### Added

- add `speed` parameter [#80](https://github.com/SpringQL/replayman/pull/80)

## [v0.2.1] - 2022-11-04

Started writing the CHANGELOG.

---

<!-- Links -->
[Keep a Changelog]: https://keepachangelog.com/
[Semantic Versioning]: https://semver.org/

<!-- Versions -->
[Unreleased]: https://github.com/SpringQL/replayman/compare/v0.5.1...HEAD
[Released]: https://github.com/SpringQL/replayman/releases
[v0.5.1]: https://github.com/SpringQL/replayman/compare/v0.5.0...v0.5.1
[v0.5.0]: https://github.com/SpringQL/replayman/compare/v0.4.0...v0.5.0
[v0.4.0]: https://github.com/SpringQL/replayman/compare/v0.3.0...v0.4.0
[v0.3.0]: https://github.com/SpringQL/replayman/compare/v0.2.1...v0.3.0
[v0.2.1]: https://github.com/SpringQL/replayman/releases/v0.2.1


src/cmd_parser.rs
use std::net::ToSocketAddrs;

use anyhow::{anyhow, Context, Result};
use clap::Parser;
use time::OffsetDateTime;

use crate::{
    destination::Destination,
    timed_stream::{file_type::FileType, TimedStream},
};

// This file is part of https://github.com/SpringQL/replayman which is licensed under MIT OR Apache-2.0. See file LICENSE-MIT or LICENSE-APACHE for full license details.

/// Log agent to replay time-stamped log stream.
#[derive(Parser)]
#[clap(author = "Sho Nakatani <lay.sakura@gmail.com>", version)]
struct Opts {
    /// Either of: tsv
    #[clap(long, default_value = "tsv")]
    log_file_type: String,

    /// Field name of log file to be used as timestamp.
    /// Field value must be formatted in RFC-3339 (e.g. 2006-04-13T14:12:53.4242+05:30)
    #[clap(long)]
    timed_by: String,

    /// (Virtual) timestamp to start replay from.
    /// Must be formatted in RFC-3339 (e.g. 2006-04-13T14:12:53.4242+05:30)
    #[clap(long)]
    initial_timestamp: String,

    /// Playback speed.
    /// If you set it to 0.5, for example, the internal virtual timer ticks at half the speed compared to wall-clock timers.
    #[clap(long, default_value = "1")]
    speed: String,

    /// TCP address:port to write logs to.
    ///
    /// (e.g. --dest-tcp 'localhost:19870')
    #[clap(long)]
    dest_tcp: Option<String>,

    /// MQTT address:port to publish logs to.
    ///
    /// (e.g. --dest-mqtt 'localhost:19870' --dest-mqtt-topic 'your/topic')
    #[clap(long)]
    dest_mqtt: Option<String>,
    /// MQTT topic.
    #[clap(long)]
    dest_mqtt_topic: Option<String>,

    /// Kafka bootstrap server's address:port to publish logs to.
    ///
    /// (e.g. --dest-kafka 'localhost:9092' --dest-kafka-topic 'your-topic')
    #[clap(long)]
    dest_kafka_bootstrap: Option<String>,
    /// Kafka topic.
    #[clap(long)]
    dest_kafka_topic: Option<String>,

    /// Log file to replay
    log_file_path: String,
}

pub(super) struct CmdParser(Opts);

impl CmdParser {
    pub(super) fn new() -> Self {
        let opts: Opts = Opts::parse();
        Self(opts)
    }

    pub(super) fn logs(&self) -> Result<TimedStream> {
        let log_file_type = match self.0.log_file_type.as_str() {
            "tsv" => Ok(FileType::Tsv),
            _ => Err(anyhow!(
                "unsupported log file type: {}",
                self.0.log_file_type
            )),
        }?;
        let timed_by = self.0.timed_by.to_string();
        let initial_timestamp = OffsetDateTime::parse(
            self.0.initial_timestamp.as_str(),
            &time::format_description::well_known::Rfc3339,
        )?;
        let speed = self.0.speed.parse::<f32>()?;
        let log_file_path = self.0.log_file_path.as_str();

        TimedStream::new(
            log_file_type,
            log_file_path,
            timed_by,
            initial_timestamp,
            speed,
        )
    }

    pub(super) fn dest(&self) -> Result<Destination> {
        match (
            &self.0.dest_tcp,
            &self.0.dest_mqtt,
            &self.0.dest_mqtt_topic,
            &self.0.dest_kafka_bootstrap,
            &self.0.dest_kafka_topic,
        ) {
            // stdout
            (None, None, None, None, None) => Ok(Destination::Stdout),

            // TCP
            (Some(tcp_addr), None, None, None, None) => {
                let addr = tcp_addr
                    .to_socket_addrs()?
                    .next()
                    .context("empty address?")?;
                Ok(Destination::Tcp(addr))
            }

            // MQTT
            (None, Some(mqtt_addr), Some(mqtt_topic), None, None) => {
                let errmsg = || format!("failed to parse MQTT address: {}", mqtt_addr);

                let mut addr = mqtt_addr.split(':');

                let host = addr.next().with_context(errmsg)?;
                let port = addr.next().with_context(errmsg)?;
                let port: u16 = port.parse().with_context(errmsg)?;

                Ok(Destination::Mqtt {
                    host: host.to_string(),
                    port,
                    topic: mqtt_topic.to_string(),
                })
            }

            // Kafka
            (None, None, None, Some(kafka_bootstrap_servers), Some(kafka_topic)) => {
                Ok(Destination::Kafka {
                    bootstrap_servers: kafka_bootstrap_servers.to_string(),
                    topic: kafka_topic.to_string(),
                })
            }

            _ => Err(anyhow!("Conflict in `--dest-*` option(s)")),
        }
    }
}
